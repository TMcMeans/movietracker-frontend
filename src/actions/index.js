export const addMovies = movieArray => {
  return {
    type: 'ADD_MOVIES',
    movies: movieArray
  };
};

export const loginUser = user => {
  return {
    type: 'LOGIN_USER',
    user
  };
};

export const logoutUser = () => {
  return {
    type: 'LOGOUT_USER'
  };
};

export const hasErrored = message => ({
  type: 'HAS_ERRORED',
  message
});

export const removeFavorites = () => ({
  type: 'REMOVE_FAVORITES'
});

export const retrieveAllFavorites = favorites => {
  return {
    type: 'RETRIEVE_ALL_FAVORITES',
    favorites
  };
};

export const addMessage = message => {
  return {
    type: 'ADD_MESSAGE',
    message
  };
};
