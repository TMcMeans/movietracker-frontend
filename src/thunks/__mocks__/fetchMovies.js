export const fetchMovies = jest.fn();
