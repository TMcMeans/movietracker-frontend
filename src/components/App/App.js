import React, { Component } from 'react';
import { connect } from 'react-redux';
import { key } from '../../apikey';
import * as API from '../../helpers/apiCalls';
import { addMovies } from '../../actions';
import CardContainer from '../../containers/CardContainer/CardContainer';
import Menu from '../../components/Menu/Menu'
import Login from '../../containers/Login/Login';
import '../../styles/main.scss';

import './App.css';

class App extends Component {
  async componentDidMount() {
    const url = `https://api.themoviedb.org/3/discover/movie?api_key=${key}&language=en-US&sort_by=revenue.desc&include_adult=false&include_video=false&page=1&with_genres=27&without_genres=10751`;
    const movies = await API.fetchMovies(url);
    this.props.addMoviesToStore(movies);
  }

  render() {
    return (
      <div className="App">
        <h1 className="header">Movie Stalker</h1>
        <Menu />
        {/* <Login /> */}
        <CardContainer />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addMoviesToStore: moviesArray => dispatch(addMovies(moviesArray))
});

export default connect(
  null,
  mapDispatchToProps
)(App);
