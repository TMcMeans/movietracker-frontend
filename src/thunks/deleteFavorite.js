import { hasErrored } from '../actions';

export const deleteFavorite = (userId, movieId) => {
  return async (dispatch) => {
    try {
      const response = await fetch(`http://localhost:3000/api/users/${userId}/favorites/${movieId}`, {
        method: 'DELETE',
        headers: {
          'Content-Type': 'application/json'
        }
      });
      if (!response.ok) {
        throw Error(response.statusText);
      }
      console.log(response)
    } catch(err) {
      dispatch(hasErrored(err.message));
    }
  }
}