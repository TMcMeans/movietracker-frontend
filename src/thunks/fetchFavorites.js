import { hasErrored, retrieveAllFavorites } from "../actions";

export const fetchFavorites = (userId) => {
  return async(dispatch) => {
    try {
      const response = await fetch(`http://localhost:3000/api/users/${userId}/favorites`);
      if (!response.ok) {
        throw Error(response.statusText)
      }
      const result = await response.json()
      console.log(result)
      dispatch(retrieveAllFavorites(result.data))
    } catch(err) {
      dispatch(hasErrored(err.message))
    }
  }
}