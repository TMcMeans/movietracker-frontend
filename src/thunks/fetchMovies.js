import { addMovies, hasErrored } from '../actions';

export const fetchMovies = url => {
  return async dispatch => {
    try {
      const response = await fetch(url);
      if (!response.ok) {
        console.log(response)
        throw Error(response.statusText);
      }
      const movies = await response.json();
      dispatch(addMovies(movies.results));
    } catch (error) {
      dispatch(hasErrored(error.message));
    }
  };
};