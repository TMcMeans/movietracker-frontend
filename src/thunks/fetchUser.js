import { hasErrored, loginUser, addMessage } from "../actions";

export const fetchUser = (email, password) => {
  return async (dispatch) => {
    try {
      const response = await fetch('http://localhost:3000/api/users', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          email,
          password
        })
      });
      if(!response.ok) {
        dispatch(addMessage('Password or Email does not Match'))
        throw Error(response.statusText)
      }
      const result = await response.json()
      dispatch(loginUser({name: result.data.name, id: result.data.id}))
      dispatch(addMessage('Sucess! You are now Logged in'))
    } catch(err) {
      console.log(err)
      dispatch(hasErrored(err.message))
      dispatch(addMessage('Internal Server Error, Failed to Login'))
    }
  }
}