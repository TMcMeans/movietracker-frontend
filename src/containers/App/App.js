import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Switch, Route, Redirect, withRouter } from 'react-router';
import PropTypes from 'prop-types';
import { key } from '../../apikey';
import CardContainer from '../../containers/CardContainer/CardContainer';
import Menu from '../../components/Menu/Menu';
import Login from '../../containers/Login/Login';
import { Favorites } from '../../components/Favorites/Favorites';
import { Link } from 'react-router-dom';
import '../../styles/main.scss';
import { fetchMovies } from '../../thunks/fetchMovies';

export class App extends Component {
  async componentDidMount() {
    const url = `https://api.themoviedb.org/3/discover/movie?api_key=${key}&language=en-US&sort_by=revenue.desc&include_adult=false&include_video=false&page=1&with_genres=27&without_genres=10751`;
    this.props.fetchMovies(url);
  }

  render() {
    return (
      <div className="App">
        <h1 className="header">
          <span className="movie">
            <Link to="/">Movie</Link>
          </span>
          <span className="stalker">
            <Link to="/">Stalker</Link>
          </span>
        </h1>
        <Menu />
        <Switch>
          <Route exact path="/" component={CardContainer} />
          <Route
            path="/login"
            render={({ match }) => {
              return <Login />;
            }}
          />
          <Route path="/favorites" component={Favorites} />
          <Redirect to="/" />
        </Switch>
      </div>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  fetchMovies: url => dispatch(fetchMovies(url))
});

export default withRouter(
  connect(
    null,
    mapDispatchToProps
  )(App)
);

App.propTypes = {
  fetchMovies: PropTypes.func.isRequired
};
