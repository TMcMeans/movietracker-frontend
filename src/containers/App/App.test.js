import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { key } from '../../apikey';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import { fetchMovies } from '../../thunks/fetchMovies';
import { fetchFavorites } from '../../thunks/fetchFavorites';
import { loginUser } from '../../actions';
import LocalStorageMock from './localStorageMock';

jest.mock('../../thunks/fetchMovies');
jest.mock('../../thunks/fetchFavorites');

describe('App', () => {
  const wrapper = shallow(
    <App fetchMovies={jest.fn()} addFavoritesToStore={jest.fn()} />
  );
  it('matches snapshot with all data correctly rendered', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call fetchMovies when the component mounts', () => {
    const expected = `https://api.themoviedb.org/3/discover/movie?api_key=${key}&language=en-US&sort_by=revenue.desc&include_adult=false&include_video=false&page=1&with_genres=27&without_genres=10751`;
    const wrapper = shallow(<App fetchMovies={jest.fn()} addFavoritesToStore={jest.fn()} />);
    const { fetchMovies } = wrapper.instance().props
    
    expect(fetchMovies).toHaveBeenCalledWith(expected);
  })

  it('should call addFavoritesToStore if there is a user logged in', () => {
    const mockUser = {name: 'Chucky', id: 2}
    const wrapper = shallow(<App addFavoritesToStore={jest.fn()} user={null} fetchMovies={jest.fn()} />)
    
    expect(wrapper.instance().props.user).toEqual(null);
    wrapper.setProps({
      user: mockUser
    })

    expect(wrapper.instance().props.addFavoritesToStore).toHaveBeenCalledWith(2);
  });

  describe('checkLocalStorage', () => {
    it('should pull a user from local storage and call addStoredUser with the user object', () => {
      const mockUser = {name: 'Jason V', id: 1}
      
      window.localStorage = new LocalStorageMock();
      localStorage.setItem('user', JSON.stringify(mockUser))

      const wrapper = shallow(
      <App 
        addFavoritesToStore={jest.fn()} 
        user={null} 
        fetchMovies={jest.fn()} 
      />);

      wrapper.setProps({addStoredUser: jest.fn()});
      const { addStoredUser } = wrapper.instance().props;
      wrapper.instance().checkLocalStorage();
      
      expect(addStoredUser).toHaveBeenCalledWith(mockUser);
    });
  });


  describe('mapStateToProps', () => {
    it('should return a user object', () => {
      const mockState = {
        movies: [
          { title: 'Halloween' },
          { title: 'It' },
          { title: 'The Conjuring' }
        ],
        user: { name: 'Tanjie', id: 4 },
        error: '',
        favorites: [
          { title: 'Halloween' },
          { title: 'It' },
          { title: 'The Conjuring' }
        ],
        message: null
      };

      const expected = { name: 'Tanjie', id: 4 };

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps.user).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    const mockDispatch = jest.fn();
    const url = 'someplace.com';

    it('should dispatch fetchMovies thunk when fetchMovies is called from props', () => {
      fetchMovies.mockImplementation(() => {});

      const expected = fetchMovies(url);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchMovies(url);

      expect(mockDispatch).toHaveBeenCalledWith(expected);
    })

    it('should dispatch fetchFavorites thunk when addFavoritesToStore is called from props', () => {
      fetchFavorites.mockImplementation(() => {})
      const expected = fetchFavorites(2);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addFavoritesToStore(2);

      expect(mockDispatch).toHaveBeenCalledWith(expected);
    });

    it('should dispatch loginUser action when addStoredUser is called', () => {
      const mockUser = {name: "Bride Of Chucky", id: 4}
      const expected = loginUser(mockUser);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addStoredUser(mockUser);

      expect(mockDispatch).toHaveBeenCalledWith(expected);
    })
  });
});