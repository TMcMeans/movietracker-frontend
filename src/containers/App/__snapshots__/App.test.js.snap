// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot with all data correctly rendered 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    fetchMovies={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="header"
        >
          <span
            className="movie"
          >
            <Link
              replace={false}
              to="/"
            >
              Movie
            </Link>
          </span>
          <span
            className="stalker"
          >
            <Link
              replace={false}
              to="/"
            >
              Stalker
            </Link>
          </span>
        </h1>,
        <Connect(Menu) />,
        <Switch>
          <Route
            component={[Function]}
            exact={true}
            path="/"
          />
          <Route
            path="/login"
            render={[Function]}
          />
          <Route
            component={[Function]}
            path="/favorites"
          />
          <Redirect
            push={false}
            to="/"
          />
        </Switch>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="movie"
            >
              <Link
                replace={false}
                to="/"
              >
                Movie
              </Link>
            </span>,
            <span
              className="stalker"
            >
              <Link
                replace={false}
                to="/"
              >
                Stalker
              </Link>
            </span>,
          ],
          "className": "header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Link
                replace={false}
                to="/"
              >
                Movie
              </Link>,
              "className": "movie",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Movie",
                "replace": false,
                "to": "/",
              },
              "ref": null,
              "rendered": "Movie",
              "type": [Function],
            },
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Link
                replace={false}
                to="/"
              >
                Stalker
              </Link>,
              "className": "stalker",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Stalker",
                "replace": false,
                "to": "/",
              },
              "ref": null,
              "rendered": "Stalker",
              "type": [Function],
            },
            "type": "span",
          },
        ],
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Route
              component={[Function]}
              exact={true}
              path="/"
            />,
            <Route
              path="/login"
              render={[Function]}
            />,
            <Route
              component={[Function]}
              path="/favorites"
            />,
            <Redirect
              push={false}
              to="/"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "path": "/login",
              "render": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "path": "/favorites",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "push": false,
              "to": "/",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="header"
          >
            <span
              className="movie"
            >
              <Link
                replace={false}
                to="/"
              >
                Movie
              </Link>
            </span>
            <span
              className="stalker"
            >
              <Link
                replace={false}
                to="/"
              >
                Stalker
              </Link>
            </span>
          </h1>,
          <Connect(Menu) />,
          <Switch>
            <Route
              component={[Function]}
              exact={true}
              path="/"
            />
            <Route
              path="/login"
              render={[Function]}
            />
            <Route
              component={[Function]}
              path="/favorites"
            />
            <Redirect
              push={false}
              to="/"
            />
          </Switch>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="movie"
              >
                <Link
                  replace={false}
                  to="/"
                >
                  Movie
                </Link>
              </span>,
              <span
                className="stalker"
              >
                <Link
                  replace={false}
                  to="/"
                >
                  Stalker
                </Link>
              </span>,
            ],
            "className": "header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Link
                  replace={false}
                  to="/"
                >
                  Movie
                </Link>,
                "className": "movie",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Movie",
                  "replace": false,
                  "to": "/",
                },
                "ref": null,
                "rendered": "Movie",
                "type": [Function],
              },
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Link
                  replace={false}
                  to="/"
                >
                  Stalker
                </Link>,
                "className": "stalker",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Stalker",
                  "replace": false,
                  "to": "/",
                },
                "ref": null,
                "rendered": "Stalker",
                "type": [Function],
              },
              "type": "span",
            },
          ],
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Route
                component={[Function]}
                exact={true}
                path="/"
              />,
              <Route
                path="/login"
                render={[Function]}
              />,
              <Route
                component={[Function]}
                path="/favorites"
              />,
              <Redirect
                push={false}
                to="/"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
                "exact": true,
                "path": "/",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "path": "/login",
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
                "path": "/favorites",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "push": false,
                "to": "/",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
